src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp

```diff
@@ -60,46 +62,64 @@ struct InitializeM1Tags {
-        tmpl::at<typename Metavariables::factory_creation::factory_classes,
-                 evolution::initial_data::InitialData>;
-    call_with_dynamic_type<void, derived_classes>(
-        &db::get<evolution::initial_data::Tags::InitialData>(box),
-        [&box](const auto* const data_or_solution) {
-          static constexpr size_t dim = System::volume_dim;
-          const double initial_time = db::get<::Tags::Time>(box);
-          const size_t num_grid_points =
-              db::get<domain::Tags::Mesh<dim>>(box).number_of_grid_points();
-          const auto& inertial_coords =
-              db::get<domain::Tags::Coordinates<dim, Frame::Inertial>>(box);
-
-          db::mutate<evolved_variables_tag>(
-              [initial_time, &inertial_coords, &data_or_solution](
-                  const gsl::not_null<EvolvedVars*> evolved_vars) {
-                evolved_vars->assign_subset(
-                    evolution::Initialization::initial_data(
-                        *data_or_solution, inertial_coords, initial_time,
-                        typename evolved_variables_tag::tags_list{}));
-              },
-              make_not_null(&box));
-
-          // Get hydro variables
-          HydroVars hydro_variables{num_grid_points};
-          hydro_variables.assign_subset(evolution::Initialization::initial_data(
-              *data_or_solution, inertial_coords, initial_time,
-              typename hydro_variables_tag::tags_list{}));
-
-          M1Vars m1_variables{num_grid_points, -1.};
-          Initialization::mutate_assign<simple_tags>(make_not_null(&box),
-                                                     std::move(hydro_variables),
-                                                     std::move(m1_variables));
+
+    static constexpr size_t dim = System::volume_dim;
+    const double initial_time = db::get<::Tags::Time>(box);
+    const size_t num_grid_points =
+        db::get<domain::Tags::Mesh<dim>>(box).number_of_grid_points();
+    const auto& inertial_coords =
+        db::get<domain::Tags::Coordinates<dim, Frame::Inertial>>(box);
+
+    using initial_data_evo_tags = typename evolved_variables_tag::tags_list;
+    using initial_data_hydro_tags = typename hydro_variables_tag::tags_list;
+
+    using initial_data_list =
+        tmpl::append<RadiationTransport::M1Grey::AnalyticData::all_data,
+                     RadiationTransport::M1Grey::Solutions::all_solutions>;
+
+    const auto initial_data_evo_vars = call_with_dynamic_type<
+        tuples::tagged_tuple_from_typelist<initial_data_evo_tags>,
+        initial_data_list>(
+        &Parallel::get<evolution::initial_data::Tags::InitialData>(cache),
+        [&inertial_coords, &initial_time](const auto* const initial_data) {
+          return evolution::Initialization::initial_data(
+              *initial_data, inertial_coords, initial_time,
+              initial_data_evo_tags{});
+        });
+
+    const auto initial_data_hydro_vars = call_with_dynamic_type<
```

Comment:
Combine the two calls to `call_with_dynamic_type`, just because I suspect it may be expensive to compile.  You can return a pair and use structured bindings to keep the variables as they are now.

Reply:
Addressed.  This way only calls `call_with_dynamic_type` once as well.

Reply:
Credit to @nilsvu for the help